base_rest_pydantic/restapi.py

```diff
@@ -0,0 +1,195 @@
+        try:
+            return self._model_cls(**params)
+        except ValidationError as ve:
+            raise UserError(_("BadRequest %s") % ve.json(indent=0))
+
+    def to_response(self, service, result):
+        # do we really need to validate the instance????
+        json_dict = result.dict()
+        to_validate = (
+            json_dict if not result.__config__.orm_mode else result.dict(by_alias=True)
+        )
+        *_, validation_error = validate_model(self._model_cls, to_validate)
+        if validation_error:
+            raise SystemError(_("Invalid Response %s") % validation_error)
+        return json_dict
+
+    def to_openapi_query_parameters(self, servic, spec):
+        json_schema = self._model_cls.schema()
+        parameters = []
+        for prop, spec in list(json_schema["properties"].items()):
+            params = {
+                "name": prop,
+                "in": "query",
+                "required": prop in json_schema.get("required", []),
+                "allowEmptyValue": spec.get("nullable", False),
+                "default": spec.get("default"),
+            }
+            if spec.get("schema"):
+                params["schema"] = spec.get("schema")
+            else:
+                params["schema"] = {"type": spec["type"]}
+            if spec.get("items"):
+                params["schema"]["items"] = spec.get("items")
+            if "enum" in spec:
+                params["schema"]["enum"] = spec["enum"]
+
+            parameters.append(params)
+
+            if spec["type"] == "array":
+                # To correctly handle array into the url query string,
+                # the name must ends with []
+                params["name"] = params["name"] + "[]"
+
+        return parameters
+
+    # TODO, we should probably get the spec as parameters. That should
+    # allows to add the definition of a schema only once into the specs
+    # and use a reference to the schema into the parameters
+    def to_openapi_requestbody(self, service, spec):
+        return {"content": {"application/json": {"schema": self.to_json_schema(spec)}}}
+
+    def to_openapi_responses(self, service, spec):
+        return {
+            "200": {
+                "content": {"application/json": {"schema": self.to_json_schema(spec)}}
+            }
+        }
+
+    def to_json_schema(self, spec):
+        schema = self._model_cls.schema()
+        schema_name = schema["title"]
+        if schema_name not in spec.components.schemas:
+            definitions = schema.get("definitions", {})
+            for name, sch in definitions.items():
+                if name in spec.components.schemas:
+                    continue
+                spec.components.schema(name, sch)
+            schema = replace_ref_in_schema(schema, schema)
+            spec.components.schema(schema_name, schema)
+        return {"$ref": f"#/components/schemas/{schema_name}"}
+
+
+class PydanticModelList(PydanticModel):
```

Comment:
@sbidoul I wonder if I should delete this class.  Lists could be supported by the base class by parsing the type given as a parameter and the specific list processing arguments only used in the case of a list.
```python
from typing import List

...


    @restapi.method(
        [(["/", "/search"], "GET")],
        input_param=PydanticModel(PartnerSearchParam),
        output_param=PydanticModel(List[PartnerShortInfo]),
        auth="public",
    )
    def search(self, partner_search_param):
        """
        """
```

Reply:
Do we actually need PydanticModel at all in the input/output param declaration ? Could it be

```
input_param=PartnerSearchParam, 
output_param=List[PartnerShortInfo],
```

and let the framework detect they derive from pydantic.BaseModel to do its magic ?

Reply:
iow, PydanticModel and PydanticModelList could be internal to the base_rest framework with `restapi.method` wrapping input_param and output_param in PydanticModel when it detects they derive from pydantic.BaseModel.

Reply:
for sure...

Reply:
I don't want to add more work on you. It can come as a future ergonomy improvement.

Reply:
I'll leave the code as is. It will be part of the next big evolution.